bolt/lib/Profile/StaleProfileMatching.cpp

```diff
@@ -45,6 +45,8 @@ namespace opts {
 
 extern cl::opt<bool> TimeRewrite;
 extern cl::OptionCategory BoltOptCategory;
+extern cl::opt<unsigned> Verbosity;
+extern cl::opt<bool> ProfileUsePseudoProbes;
```

Comment:
IIUC `ProfileUsePseudoProbes` controls whether we extract probe from binary and write it out in YAML profile today. It might be confusing to repurpose this same flag to control stale profile matching strategy. Something specific for profile matching strategy might be better.

Reply:
No, I intended `ProfileUsePseudoProbes` to cover both generation and use of pseudo probes. The option means we parse pseudo probes for use in Profile, and that covers both YAMLProfileReader and Writer.

Reply:
Ok, I think we need a dedicated flag to control matching strategy, otherwise it can be confusing. 

Generating profile with probe and using profile for stale profile matching are two different things. Like profile contains probe doesn't mean stale matching has to use probe.

Reply:
> Ok, I think we need a dedicated flag to control matching strategy, otherwise it can be confusing. Generating profile with probe and using profile for stale profile matching are two different things.

Agree they're different but both require parsing pseudo probes, which is controlled by the flag. `ProfileUseDFS` is another flag which controls both profile generation and consumption.

> Like profile contains probe doesn't mean stale matching has to use probe.

If the profile contains probe information, but ProfileUsePseudoProbes is not set, we will not use the information for matching.

Reply:
> If the profile contains probe information, but ProfileUsePseudoProbes is not set, we will not use the information for matching.

But this is the confusion... The profile (not stale matching) is clearly using probe as probe exists in profile, why you want users to use such a flag (set to false) to control stale matching strategy. Wouldn't it be better to have a dedicated flag whose name actually reflects its purpose in this case?

Reply:
> The profile (not stale matching) is clearly using probe as probe exists in profile

But in a similar way, the profile can contain BB hashes, but they won't be used unless `-infer-stale-profile` is set. (We produce them unconditionally so there's no corresponding generation flag).

There's no use of probes outside stale matching.

The confusion argument for me is flipped – if there were separate flags for generation and consumption for probes. It's as if we had `ProfileEmitDFSIndices` and `ProfileAssumeDFSIndices` instead of `ProfileUseDFS` which covers both.

Reply:
> There's no use of probes outside stale matching. 

True, but you do have the options to let probe live in yaml profile and then not use it (even just for testing). you do have a state this is NOT "having probe and use it for profile matching". 

Wouldn't a switch like `enable-probe-based-profile-matching` reflect the intention much better? We may even need something like `profile-matching-strategy={probe,hash,call}` in the future. 

> It's as if we had ProfileEmitDFSIndices and ProfileAssumeDFSIndices instead of ProfileUseDFS which covers both.

This is different, DFS order is a property of the profile. Whether you use probe to do matching is not.

Reply:
Added a flag for pseudo probe block matching.